flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaInvocationHandler.java

```diff
@@ -205,12 +207,34 @@ private Object invokeRpc(Method method, Object[] args) throws Exception {
 			result = null;
 		} else if (Objects.equals(returnType, CompletableFuture.class)) {
 			// execute an asynchronous call
-			result = ask(rpcInvocation, futureTimeout);
+			CompletableFuture resultFuture = ask(rpcInvocation, futureTimeout);
+
+			result = resultFuture.thenApply((Object o) -> {
+				if (o instanceof SerializedValue) {
+					try {
+						return  ((SerializedValue) o).deserializeValue(getClass().getClassLoader());
+					} catch (IOException | ClassNotFoundException e) {
+						throw new CompletionException(e);
+					}
+				} else {
+					return o;
+				}
+			});
 		} else {
 			// execute a synchronous call
-			CompletableFuture<?> futureResult = ask(rpcInvocation, futureTimeout);
+			CompletableFuture<?> resultFuture = ask(rpcInvocation, futureTimeout);
 
-			result = futureResult.get(futureTimeout.getSize(), futureTimeout.getUnit());
+			Object returnedResult = resultFuture.get(futureTimeout.getSize(), futureTimeout.getUnit());
+
+			if (returnedResult instanceof SerializedValue) {
+				try {
+					result = ((SerializedValue) returnedResult).deserializeValue(getClass().getClassLoader());
+				} catch (IOException | ClassNotFoundException e) {
+					throw new CompletionException(e);
```

Comment:
This should not be a `CompletionException`. Instead I would throw a `RpcException` saying that one could not deserialize a serialized payload.